### PR TITLE
[Core] Upgrade to pytorch 2.2, remove cupy dependency, avoid nccl 2.19 bug

### PR DESCRIPTION
**Pull request by youkaichao**
<details>

<summary><b> PR Checklist (Click to expand. Please read before submitting.) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

---

vLLM will allocate all the available memory, so we need to run the tests one by one. The solution is to pass arguments (model name) by environment variables.

### CODE REVIEW

**Comment by youkaichao** [comment]
Let me test with this modification first, and then test with torch 2.2.0 . In my dev box, this fix can work with torch 2.2.0 .

**Comment by youkaichao** [comment]
I'm trying to bisect which pytorch nightly version starts to fail vllm. The most annoying thing is we have to build xformers from source each time, which takes quite a long time.

At least `2.2.0.dev20231015` is good:

```sh
# install pre-build pytorch nightly wheel
pip install https://download.pytorch.org/whl/nightly/pytorch_triton-2.1.0%2B6e4932cda8-cp39-cp39-linux_x86_64.whl
pip install https://download.pytorch.org/whl/nightly/cu121/torch-2.2.0.dev20231015%2Bcu121-cp39-cp39-linux_x86_64.whl
# verify basic cuda functionality
python -c "import torch; print(torch.randn(5, 5, 5).cuda().sum())"
# verify pt2 stack
python -c "import torch; print(torch.compile(lambda x: x + 1)(torch.randn(5, 5, 5).cuda()).sum())"

# build xformers from source
pushd ../xformers
pip install "numpy<2"
# avoid CPU overload
MAX_JOBS=4 python setup.py develop
# verify xformers
python -c "from xformers.ops import memory_efficient_attention; import torch; data = torch.rand(8, 128, 32, 128).cuda(); output = memory_efficient_attention(data, data, data); print(output.sum())"
popd

# build vllm
pip install -r requirements-build.txt
pip install -r requirements-dev.txt
python setup.py develop

# test vllm
TEST_DIST_MODEL=meta-llama/Llama-2-7b-hf python -m pytest -s tests/distributed/test_basic_distributed_correctness.py
```

**Comment by youkaichao** [comment]
Further investigation shows `torch-2.3.0.dev20231221+cu121` contains the `cuInit` bug. Starting from `torch-2.3.0.dev20231222+cu121`, this `cuInit` bug has gone. But then the OOM issue is still there.

**Review by rkooo567 — COMMENTED** [review]
I wonder if we can instead just clean up states between tests using 

```
    import gc
    import torch
    gc.collect()
    torch.cuda.empty_cache()
```
?

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +15,9 @@ set(PYTHON_SUPPORTED_VERSIONS "3.8" "3.9" "3.10" "3.11")
 # Supported NVIDIA architectures.
 set(CUDA_SUPPORTED_ARCHS "7.0;7.5;8.0;8.6;8.9;9.0")
 
+# used when building pytorch-related extensions
+set(TORCH_CUDA_ARCH_LIST "7.0;7.5;8.0;8.6;8.9;9.0")
```
Comment (youkaichao) [inline CMakeLists.txt:19]: note that pytorch 2.2.0 has 9.0a support by default:

https://github.com/pytorch/pytorch/blob/19d27a13ea052230d9fb565a5b82e683e28d1697/Dockerfile#L60

while our docker image does not support 9.0a

**Comment by youkaichao** [comment]
After this PR, we don't need `cupy_port` stuff.

The initialization of distribution is as follows:

- We use pytorch to initialize distributed environment
- `pynccl` also uses pytorch to initialize `nccl` communicators. After that, its allreduce operation directly operates on pytorch tensor.

Note: with this setup, it should be easy to scale to multi-node parallel, leveraging the `torch.distributed`.

Then we can completely remove the dependency of `cupy`. And hopefully we can have a clean logic of distributed inference.

There are many followup cleanup work to do, but I prefer to leave them as a new PR. This PR has already grown quite large.

File: vllm/model_executor/parallel_utils/pynccl.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+import ctypes
+import torch.distributed as dist
+from torch.distributed import ReduceOp
+import datetime
+import os
+import glob
+
+# manually load the nccl library
+_path = os.path.dirname(os.path.abspath(__file__))
+so_file = glob.glob(f"{_path}/../../lib/nvidia/nccl/lib/libnccl.so.*")[0]
```
Comment (zhuohan123) [inline vllm/model_executor/parallel_utils/pynccl.py:24]: Haven't looked into the whole PR. But I feel having this kind of absolute path for libnccl seems very dangerous. Many cloud providers may ship their own nccl package (e.g., aws for efa support). I'm not sure about what's the effect of this.

Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py:24]: This is our secrect version of `nccl`, living inside `vllm`, we ship it with our wheel. Please check the modification in `setup.py`.

We can remove this after `nccl` fixes the bug, or after pytorch downgrades to `nccl==2.18`.

Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py:24]: Expanding the path, you can see it is `/path/to/vllm/lib/nvidia/nccl/lib/libnccl.so.2.18.3`. I created a secrect `lib` folder under `vllm` library, and this `.so` file will come with `vllm` installation.

**Comment by rkooo567** [comment]
Imo, this is a great direction. I've seen many successful cases reducing external dependencies like this that are not necessary + I've seen lots of users having to turn off cuda graph because cupy didn't work with their environment (and it will resolve their issues)

**Comment by youkaichao** [comment]
> because cupy didn't work with their environment (and it will resolve their issues)

Thanks for your comment! I will completely remove `cupy` dependency in a followup PR.

**Review by rkooo567 — COMMENTED** [review]
Do you have a benchmark result before/after the PR?

File: setup.py [hunk 3]
```diff
@@ -188,6 +194,50 @@ def _install_punica() -> bool:
...
+if _is_cuda():
+
+    # tricky part, nccl 2.19 has a bug that increased memory overhead
+    # of cudagraph. However, pytorch has binary dependencies on nccl 2.19,
+    # simply `pip install nvidia-nccl-cu12==2.18.3` will break pytorch,
+    # so we have to manually download nccl 2.18 and keep the library to
+    #  a secrect place
+
+    # Define the URL of the file and the directory to unzip to
+    file_url = ('https://files.pythonhosted.org/packages/44/6e/'
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply setup.py:206]: Actually I wonder if we can support env var so that we can also decide to load arbitrary .so instead of always downloading our package when we build

[the remark this answers is not in the document]
Comment (youkaichao) [reply setup.py:206]: detecting `env var` during runtime is good.

w.r.t. downloading our package when we build, we have to do this because nccl brought by torch==2.2.0 does not work.

File: setup.py [hunk 4]
```diff
@@ -188,6 +194,50 @@ def _install_punica() -> bool:
...
+    with tempfile.TemporaryDirectory() as temp_dir:
+        local_zip_path = (
+            f"{temp_dir}/"
+            "nvidia_nccl_cu12-2.18.3-py3-none-manylinux1_x86_64.whl")
+        # check if the target directory exists
+        if not os.path.exists(target_dir):
+            logger.info(f'Creating target directory {target_dir} ...')
+            os.makedirs(target_dir)
+        # Check if the file is already downloaded
+        if os.path.exists(target_dir + "nvidia"):
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply setup.py:220]: We have no choice here, we cannot `break` or `return`, because it is not inside any function 👀

File: vllm/model_executor/parallel_utils/pynccl.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+class NcclUniqueId(ctypes.Structure):
+    _fields_ = [("internal", ctypes.c_byte * 128)]
+
+
+# equivalent to c declaration:
+# ncclResult_t ncclGetUniqueId(ncclUniqueId* uniqueId);
+_c_ncclGetUniqueId = nccl.ncclGetUniqueId
+_c_ncclGetUniqueId.restype = ctypes.c_int
+_c_ncclGetUniqueId.argtypes = [ctypes.POINTER(NcclUniqueId)]
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py:70]: I did this to explicitly avoid `pybind11` or any other solutions requiring additional building/compilation, to reduce binary dependencies. This code will work for any nccl version, as long as they have `ncclAllReduce` function symbol.

File: vllm/model_executor/parallel_utils/pynccl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py]: No conflict. PyTorch will load `libnccl.so.2`, while our file is `libnccl.so.2.18.3` . That's the value of this PR, to separate nccl versions used by vllm and pytorch.

File: setup.py [hunk 6]
```diff
@@ -188,6 +194,50 @@ def _install_punica() -> bool:
...
+                '3c9cd7007072f8a63dae7b5eddd1cc1525fd357377467ce3a4749b02d5ff'
+                '/nvidia_nccl_cu12-2.18.3-py3-none-manylinux1_x86_64.whl')
+
+    logger.info('Installing NVIDIA NCCL library...')
+
+    target_dir = os.path.dirname(os.path.abspath(__file__)) + "/vllm/lib/"
+    with tempfile.TemporaryDirectory() as temp_dir:
+        local_zip_path = (
+            f"{temp_dir}/"
+            "nvidia_nccl_cu12-2.18.3-py3-none-manylinux1_x86_64.whl")
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply setup.py:216]: Do you mean `arm64`? We don't need to consider `amd` here because we are under the condition `if _is_cuda():` .

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -49,7 +49,7 @@ jobs:
       matrix:
           os: ['ubuntu-20.04']
           python-version: ['3.8', '3.9', '3.10', '3.11']
-          pytorch-version: ['2.1.2']  # Must be the most recent version that meets requirements.txt.
+          pytorch-version: ['2.2.0']  # Must be the most recent version that meets requirements.txt.
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply .github/workflows/publish.yml:52]: We pin nccl version 2.18.3 with our own version of nccl, downloaded during `setup.py`.

**Comment by youkaichao** [comment]
> Do you have a benchmark result before/after the PR?

I don't have a benchmark result yet. I suppose it does not change too much, only might get some performance variation (up or down) because of version change in torch & xformers .

File: .github/workflows/publish.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply .github/workflows/publish.yml:52]: Why don't we upgrade directly to PyTorch 2.2.1? I think the latest version of xFormers is built on PyTorch 2.2.1: https://github.com/facebookresearch/xformers/releases/tag/v0.0.25

[the remark this answers is not in the document]
Comment (youkaichao) [reply .github/workflows/publish.yml:52]: pytorch 2.2.1 can be done if you like.

**Comment by youkaichao** [comment]
More support evidence for pinning `nccl` versions:

1. nccl team does not consider 30x cudagraphs before, their ideal case is no more than 10 cudagraphs, and we are lucky nccl 2.18.3 works for us: https://github.com/NVIDIA/nccl/issues/1234#issuecomment-2017546657 
2. It seems quite common for users to pin nccl versions themselves, e.g. https://github.com/NVIDIA/nccl/issues/1136 pins nccl 2.17.1 , and many other reports in https://discuss.pytorch.org/t/how-can-i-change-nccl-version-in-pytorch/143942/13 

More support evidence for removing `cupy` dependencies:

We get more and more bug reports with cupy’s TCPStore recently, e.g. https://github.com/vllm-project/vllm/issues/3057 and https://github.com/vllm-project/vllm/issues/3617

File: .github/workflows/publish.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (youkaichao) [reply .github/workflows/publish.yml:52]: Then we can move to 2.2.1 . No problem.

File: CMakeLists.txt (hunk 1 above)
Comment (WoosukKwon) [reply CMakeLists.txt:19]: Doesn't this make our build system always compile the CUDA kernels for all architectures?

If I remember correctly, we only compiled the kernels for a single architecture by detecting the equipped GPUs on the user machine (I'm not sure this is still true after we changed our build system to CMake though), to reduce the compile time. Exceptionally, we targeted all architectures when building docker images or pypi wheels.

**Comment by WoosukKwon** [comment]
Also, how does this work with AMD GPUs? As you know, AMD GPUs also use ROCm version of CuPy because of some bugs in using PyTorch RCCL with HIP graphs.

**Comment by youkaichao** [comment]
> can we use an older version of NVIDIA NCCL package [pypi.org/project/nvidia-nccl-cu12](https://pypi.org/project/nvidia-nccl-cu12/) instead of including it in our package?

No, it will break pytorch, because pytorch has binary dependency on nccl 2.19.3.

**Comment by youkaichao** [comment]
> how does it work with the vLLM PyPI package if we have an additional build flag like DOWNLOAD_NCCL , and turn it off by default, but turn it on in our CI & docker image?

Then pypi version of vllm will use nccl 2.19.3, which costs more memory and would have lower performance when using cudagraph.

**Comment by youkaichao** [comment]
> Also, how does this work with AMD GPUs? As you know, AMD GPUs also use ROCm version of CuPy because of some bugs in using PyTorch RCCL with HIP graphs.

It seems RCCL has exactly the same API as NCCL. Ideally we should be able to just use `VLLM_NCCL_SO_PATH` to point to a valid RCCL path, and it should work. However I don't have a working AMD environment, and I cannot test it.

**Comment by youkaichao** [comment]
Benchmarks:

# PT 2.1.2 [f408d05](https://github.com/vllm-project/vllm/commit/f408d05c523c25e2f638a13cb34a2dab3dcb2754) (current main)

image

# PT 2.2.0 [3d9332a](https://github.com/vllm-project/vllm/pull/3442/commits/3d9332a938931522088aacca945b1063ad2424db) (this branch)
image

# PT 2.2.1 [76f46f6](https://github.com/vllm-project/vllm/pull/3442/commits/76f46f65401d6fb897ce388fe92e219fbfd6656b) (this branch)

image

**Comment by simon-mo** [comment]
The benchmark in our CI is really just a placeholder. It doesn't run distributed (tp>1) and also running tiny models.

File: CMakeLists.txt (hunk 1 above)
Comment (youkaichao) [reply CMakeLists.txt:19]: This is used in docker image. It seems this CMake inherit the build architecture from pytorch by default, and so I have to change it (to avoid 9.0a architecture that is not supported in docker nvcc).

File: setup.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hongxiayang) [reply setup.py:206]: how does this solution scale to other backends, like AMD GPU? Does it require another secret .so?

File: vllm/model_executor/parallel_utils/pynccl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hongxiayang) [reply vllm/model_executor/parallel_utils/pynccl.py]: how does this solution scale to other backends, like AMD GPU? Does it require another secret .so?

[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py]: Our CI does not run in amd GPUs. We don't know if amd `rccl` has a similar problem. Help is needed!

**Comment by youkaichao** [comment]
This PR has been separated into #3625 and #3805
